crates/ibc/src/core/ics02_client/handler/upgrade_client.rs

```diff
@@ -35,22 +35,50 @@ where
 {
     let MsgUpgradeClient { client_id, .. } = msg;
 
-    // Read client state from the host chain store.
+    // Read the current latest client state from the host chain store.
     let old_client_state = ctx.client_state(&client_id)?;
 
+    // Check if the client is frozen.
     if old_client_state.is_frozen() {
-        return Err(ClientError::ClientFrozen { client_id }.into());
+        return Err(ContextError::ClientError(ClientError::ClientFrozen {
+            client_id,
+        }));
     }
 
-    let upgrade_client_state = ctx.decode_client_state(msg.client_state)?;
-
-    if old_client_state.latest_height() >= upgrade_client_state.latest_height() {
-        return Err(ClientError::LowUpgradeHeight {
-            upgraded_height: old_client_state.latest_height(),
-            client_height: upgrade_client_state.latest_height(),
-        }
-        .into());
-    }
+    // Read the latest consensus state from the host chain store.
+    let old_consensus_state = ctx
+        .consensus_state(&client_id, &old_client_state.latest_height())
+        .map_err(|_| ClientError::ConsensusStateNotFound {
+            client_id: client_id.clone(),
+            height: old_client_state.latest_height(),
+        })?;
+
+    let now = ctx.host_timestamp()?;
+    let duration = now
+        .duration_since(&old_consensus_state.timestamp())
+        .ok_or_else(|| ClientError::InvalidConsensusStateTimestamp {
+            time1: old_consensus_state.timestamp(),
+            time2: now,
+        })?;
+
+    // Check if the latest consensus state is within the trust period.
+    if old_client_state.expired(duration) {
+        return Err(ContextError::ClientError(
+            ClientError::HeaderNotWithinTrustPeriod {
+                latest_time: old_consensus_state.timestamp(),
+                update_time: now,
+            },
+        ));
+    };
```

Comment:
ibc-go also doesn't seem to have that check?

Reply:
This one also is the same as "Client frozen" matter. 
Dropping this enables a client that is "not within the trusting period" to restart.

Reply:
I think we should answer this question:
Is there any case that by "having a frozen client" or "exceeding the allowed trusting period" we also must not preserve its connections? (in any cost)

Reply:
#391